README.md
# Rholang for VSCode

<img src="https://github.com/tgrospic/rholang-vscode/raw/master/assets/rchain.png"/>

<!-- ![beta][beta-badge] VSCode restriction for SVG in README -->

This is Visual Studio Code extension for [**Rholang**][rholang] programming language. It has support for syntax highlighting and code evaluation with error highlighting.

> **Rholang** is a fully featured, general purpose, Turing-complete programming language built from the rho-calculus. It is a behaviorally typed, **r**-eflective, **h**-igher **o**-rder process language and the official smart contracting language of [RChain][rchain-coop]. Its purpose is to concretize fine-grained, programmatic concurrency. [RChain Architecture][arch-rholang]

Rholang is currently in active development and syntax can slightly change. Current version of the plugin follows this version of [Rho grammar][rho-bnf-origin] and available examples.

Programmers in concurrent languages such as Erlang/Elixir say that one of the hardest problems is to coordinate the names (locations) of processes. It seems that Rholang with [Namespace logic][arch-namespace-logic] looks like a great solution for coordination of resources.

With all this sweet superpowers, that comes with the Rholang compiler and type checker, it is a pleasure to write (smart) `contract`s. :lollipop:

## How to use this extension

Just **open** your favorite **`.rho`** file, **write** some exciting **Rholang** code, **save** the file and **enjoy** the results in VSCode **Output Panel**. :smile:

![Rho sample][screenshot]

In the _Output Panel_ you can see the progress when RNode instance is starting in the background. It will output the message and create notification when it's ready to evaluate the code.

The Language Server is enabled and runs locally installed RNode by default. If RNode is not installed (not in the path, in extension settings it can be switched to use Docker or set custom RNode executable location.

| Settings               | Default | Description
| ---------------------- | ------- | ------------
| Enable Language Server | `true`  | Enable loading of the Language Server and RNode when open a file.
| Enable Docker          | `false` | Run RNode with Docker.
| RNode                  | `rnode` | RNode executable (path) used by Rholang Language Server.
| RNode Docker Image     | `rchain/rnode` | Docker image (version) used by Rholang Language Server.
| Show All Output        | `false` | Show all output from RNode in Output Panel.

Currently, one RNode instance is created with one VSCode instance. For each RNode instance unique temp directory is created which holds `.rnode` data directory. If VSCode window is reloaded it will create a fresh instance of RNode also.

## Learn Rholang By Example

If you are new to Rholang I recommend to visit Joshy Orndorff's excellent introduction [Learn Rholang By Example](https://github.com/JoshOrndorff/LearnRholangByExample) for beginners and intermediate programmers.

## TODO

- add commands, completion, hovers...
- connect with the compiler (get semantic info)
- ...

## Publish

### [Visual Studio Code Marketplace](https://marketplace.visualstudio.com/vscode)   
`vsce publish -p <token>`

### [Open VSX Registry](https://open-vsx.org)  
`ovsx publish -p <token>`

## Release Notes

### 0.7.0 (March 13, 2022)
- Update syntax highlighting for Rholang 1.1 (RNode v0.13.0)

### 0.6.5 (May 30, 2021)
- Update for RNode version >0.9.25 (config changes)

### 0.6.0 (October 06, 2019)
- Switch to gRPC connection :rocket:
- Update for RNode v0.9.{14,15} (new API)

### 0.5.0 (June 13, 2019)
- Update RNode option `--allow-private-addresses`

### 0.4.0 (May 18, 2019)
- Update for RNode v0.9.5

### 0.3.0 (April 20, 2019)
- Support for RNode installed locally
- Fix v0.9.3 - error detection, cost output

### 0.2.0 (February 11, 2019)
- Docker image user settings

### 0.1.0 (February 02, 2019)
- Grammar update for Rholang Mercury release v0.8>.
- **Rholang Language Server** with error highlighting.
- Snippets for basic Rholang terms.

### 0.0.1 (January 28, 2018)
- Initial release. Syntax highlighting.

## License

[The MIT License (MIT)][license]

[rholang]: https://github.com/rchain/rchain/blob/master/docs/rholang/rholangtut.md
[releases]: https://github.com/tgrospic/rholang-vscode/releases
[rchain-coop]: https://www.rchain.coop
[rho-github]: https://github.com/rchain/rchain/tree/master/rholang
[rho-bnf-origin]: https://github.com/rchain/rchain/blob/243a3fd7b1d7f1b9669920da80f419e84fbd02b6/rholang/src/main/bnfc/rholang_mercury.cf
[arch-rholang]: http://rchain-architecture.readthedocs.io/en/latest/contracts/contract-design.html#rholang-a-concurrent-language
[arch-namespace-logic]: http://rchain-architecture.readthedocs.io/en/latest/contracts/namespaces.html#namespace-logic
[tuplespaces-to-picalculus]: http://mobile-process-calculi-for-programming-the-new-blockchain.readthedocs.io/en/latest/actors-tuples-and-pi.html#from-tuplespaces-to-calculus

[beta-badge]: https://cdn.jsdelivr.net/gh/tgrospic/rholang-vscode@v0.6.1/assets/beta.svg
[screenshot]: https://github.com/tgrospic/rholang-vscode/raw/v0.3.0/assets/rholang-sample.gif
[license]: https://github.com/tgrospic/rholang-vscode/blob/master/LICENSE


CHANGELOG.md
# Change Log

### 0.7.0 (March 13, 2022)
- Update syntax highlighting for Rholang 1.1 (RNode v0.13.0)

### 0.6.5 (May 30, 2021)
- Update for RNode version >0.9.25 (config changes)

### 0.6.0 (October 06, 2019)
- Switch to gRPC connection :rocket:
- Update for RNode v0.9.{14,15} (new API)

### 0.5.0 (June 13, 2019)
- Update RNode option `--allow-private-addresses`

### 0.4.0 (May 18, 2019)
- Update for RNode v0.9.5

### 0.3.0 (April 20, 2019)
- Support for RNode installed locally
- Fix v0.9.3 - error detection, cost output

### 0.2.0 (February 11, 2019)
- Docker image user settings (default `rchain/rnode:master`)

### 0.1.0 (February 02, 2019)
- Grammar update for Rholang Mercury release v0.8>.
- **Rholang Language Server** with error highlighting.
- Snippets for basic Rholang terms.

### 0.0.1 (January 28, 2018)
- Initial release. Syntax highlighting.


src/extension.ts
'use strict'

import * as path from 'path'

import { workspace, ExtensionContext } from 'vscode'
import { LanguageClient, LanguageClientOptions, ServerOptions, TransportKind } from 'vscode-languageclient'

export function activate(context: ExtensionContext) {
  // The Rholang Language Server path
  const serverModule = context.asAbsolutePath(path.join('build/src', 'server.js'))
  // The debug options for the server
  const debugOptions = { execArgv: ["--nolazy", "--inspect=6009"] }

  // If the extension is launch in debug mode the debug server options are use
  // Otherwise the run options are used
  const serverOptions: ServerOptions = {
    // run: { module: serverModule, transport: TransportKind.stdio },
    // debug: { module: serverModule, transport: TransportKind.stdio, options: debugOptions }
    run: { module: serverModule, transport: TransportKind.ipc },
    debug: { module: serverModule, transport: TransportKind.ipc, options: debugOptions },
  }

  // Options to control the language client
  const clientOptions: LanguageClientOptions = {
    // Register the server for Rholang
    documentSelector: ['rho'],
    outputChannelName: 'Rholang',
    synchronize: {
      // Synchronize the setting section 'rholang' to the server
      configurationSection: 'rholang',
      // Notify the server about file changes to '.rholangrc files contain in the workspace
      // fileEvents: workspace.createFileSystemWatcher('**/.rholangrc')
    }
  }

  // Create the language client and start the client.
  const disposable = new LanguageClient('rho', 'Rholang Language Server', serverOptions, clientOptions).start()

  // Push the disposable to the context's subscriptions so that the
  // client can be deactivated on extension deactivation
  context.subscriptions.push(disposable)
}


src/rholang-server.ts
import Uri from 'vscode-uri'
import { IConnection, TextDocuments } from 'vscode-languageserver'
import { DiagnosticSeverity } from "vscode-languageserver-types"
import { spawn, ChildProcess } from 'child_process'
import * as R from 'ramda'
import * as fs from 'fs-extra'
import * as os from 'os'
import * as crypto from 'crypto'
import * as path from 'path'
import * as yaml from 'yaml'
// RNode gRPC
import * as grpc from '@grpc/grpc-js'
import { rnodeProtobuf, EvalRequest, ServerReflectionRequest, ServerReflectionResponse } from '@tgrospic/rnode-grpc-js'
import * as protoSchema from '../rnode-grpc-gen/js/pbjs_generated.json'
import '../rnode-grpc-gen/js/repl_pb'
import '../rnode-grpc-gen/js/reflection_pb'

// Repl gRPC settings
const rnodePort0 = 30400
const rnodeUrl = `localhost:${rnodePort0+2}`
const pbTypes = rnodeProtobuf({ protoSchema })

// Directory paths
const rhoTmpPath = path.join(os.tmpdir(), 'vscode-rholang')
const rhoTmpName = `eval-${tmpFileName().substr(0, 16)}`
const workingFolder = path.join(rhoTmpPath, rhoTmpName)
const dataDir       = path.join(workingFolder, '.rnode')

function tmpFileName() {
  const hash = crypto.createHash('sha256')
  hash.write(Math.random().toString())
  hash.end()
  return hash.read().toString('hex')
}

function regexec(regex, str) {
  let r, result, p
  r = new RegExp(regex)
  result = []
  for (;;) {
    if (p = r.exec(str)) { result.push(p) } else break
  }
  return result[0]
}

const rnodeGetReplService = (client: grpc.Client) =>
  new Promise((resolve, reject) => {
    // Call RNode reflection method
    const comm = client.makeBidiStreamRequest<ServerReflectionRequest, ServerReflectionResponse>(
      `/grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo`,
      R.pipe(pbTypes.ServerReflectionRequest.serialize, Buffer.from),
      pbTypes.ServerReflectionResponse.deserialize,
    )
    // Send request
    comm.write({ listServices: '-' })

    // Receive response
    comm.on('data', (resultMsg: ServerReflectionResponse) => {
      const services = resultMsg.listServicesResponse.serviceList
      const replService = R.find(({name}) => !!name.match(/.Repl$/i), services)
      resolve(replService)
      comm.cancel()
    })
    comm.on('error', reject)
  })

/**
 * Evaluate Rholang code on RNode
 * - handle both versions (namespaces) of API
 */
const rnodeReplEval = async (replReq: EvalRequest, replServiceName: string, client: grpc.Client) =>
  new Promise((resolve, reject) => {
    // Call RNode Eval method
    client.makeUnaryRequest(
      `/${replServiceName}/Eval`,
      R.pipe(pbTypes.EvalRequest.serialize, Buffer.from),
      pbTypes.ReplResponse.deserialize,
      replReq,
      (err, result) => {
        if (err) reject(err)
        else resolve(result)
      }
    )
  })

type Settings = {
  enableLanguageServer: boolean
  rnode               : string
  enableDocker        : boolean
  rnodeDockerImage    : string
  dockerRunOptions    : string
  showAllOutput       : boolean
  // Experimental - modify Rholang code
  unsupportedNamesRemoveEnable: boolean
  unsupportedNames            : string
}

export class RholangServer {
  protected workspaceRoot: Uri | undefined
  protected readonly documents = new TextDocuments()
  protected readonly pendingValidationRequests = new Map<string, number>()
  private _uri: string | undefined
  private _settings: Settings
  isReady = false
  grpcReplServicePath = null
  grpcClient: grpc.Client = null

  constructor(
    protected readonly connection: IConnection,
    readonly log = (msg: any) => connection.console.log(msg)
  ) {
    this.documents.listen(this.connection)

    let rhovm : ChildProcess

    const stopRNodeProcess = useDocker => {
      log(`Stopping RChain Node.`)

      if (rhovm) {
        this.log(`Process PID: ${rhovm.pid}`)

        try {
          if (useDocker) spawn('docker', ['stop', rhoTmpName], { detached: true, stdio: 'ignore' })
          else process.kill(rhovm.pid)
        } catch {}
        rhovm = null
      }
      this.isReady = false
    }

    const startRNodeProcess = () => {
      log(`Temp folder: ${workingFolder}`)

      // Start RNode (standalone) process used by the server
      rhovm = this.startRNode()
      if (rhovm) {
        this.log(`Process PID: ${rhovm.pid}`)

        // Try to stop RNode on server exit
        process.on('exit', () => stopRNodeProcess(this.useDocker()))
      }
    }

    this.connection.onInitialize(params => {
      if (params.rootPath) {
        this.workspaceRoot = Uri.file(params.rootPath)
      } else if (params.rootUri) {
        this.workspaceRoot = Uri.parse(params.rootUri)
      }
      return {
        capabilities: {
          textDocumentSync: this.documents.syncKind,
          hoverProvider: false,
          documentSymbolProvider: false,
          documentRangeFormattingProvider: false,
        }
      }
    })

    // Evaluate the code on save document
    this.connection.onDidSaveTextDocument(params => {
      const uri = params.textDocument.uri
      if(this.isReady) {
        this.evalCode(uri)
      } else if(this._settings.enableLanguageServer) {
          this.connection.window.showWarningMessage('RNode is starting, please wait...')
      }
    })

    // Settings
    let startDelayHandle : NodeJS.Timeout
    this.connection.onDidChangeConfiguration((change) => {
      this.log('Settings.' + yaml.stringify(change.settings))

      const oldSettings = this._settings
      this._settings    = change.settings.rholang || {}
      const diffs = (newS: Settings, oldS: Settings) => [
        [true              , newS.enableLanguageServer, oldS.enableLanguageServer],
        [true              , newS.enableDocker        , oldS.enableDocker],
        [true              , newS.dockerRunOptions    , oldS.dockerRunOptions],
        [!newS.enableDocker, newS.rnode               , oldS.rnode]  ,
        [newS.enableDocker , newS.rnodeDockerImage    , oldS.rnodeDockerImage],
      ]
      if(!oldSettings || diffs(this._settings, oldSettings).find(([s, x, y]) => s && x !== y)) {
        // If settings change
        oldSettings && 
          (oldSettings.enableDocker || oldSettings.dockerRunOptions) && 
          stopRNodeProcess(oldSettings.enableDocker)

        const startNode = () => {
          if (this._settings.enableLanguageServer) {
            log('Rholang Language Server (enabled)')
            startRNodeProcess()
          } else {
            log('Rholang Language Server (disabled)')
          }
        }
        // Small delay for rnode to die
        clearTimeout(startDelayHandle)
        startDelayHandle = setTimeout(startNode, 1500)
      }
    })

  }

  useDocker() { return this._settings.enableDocker }

  start() {
    this.connection.listen()
  }

  sendErrors(uri, ...msgs) {
    const diagnostics = msgs.map(x => ({
      severity: DiagnosticSeverity.Error,
      ...x
    }))
    this._settings.showAllOutput &&
      this.log('Diagnostics:\n' + yaml.stringify(diagnostics))
    this.connection.sendDiagnostics({ uri, diagnostics })
  }

  sendClear(uri) {
    const diagnostics = []
    this.connection.sendDiagnostics({ uri, diagnostics })
  }

  startRNode() {
    // Create working folder, as current user (Docker run as root)
    if (!fs.existsSync(workingFolder)) fs.ensureDir(workingFolder)

    // RNode ports
    const [port0, port1, port2, port3, port4] = [0, 1, 2, 3, 4].map(x => rnodePort0 + x)

    const runCmd = dir => [
      '--grpc-port', `${port1}`, '--grpc-port-internal', `${port2}`,
      'run', '-s', '-n', '--data-dir', dir, '--host', 'localhost', '--allow-private-addresses',
      '-p', `${port0}`, '--http-port', `${port3}`, '--kademlia-port', `${port4}`,
    ]

    const dockerCmd = runOptions => {
      const volume = `${workingFolder}:/vscode`
      let cmd = [
        // Docker run
        'run', '-i', '--rm', '--name', rhoTmpName, '-v', volume,
               '-p', `${port0}:${port0}`, '-p', `${port1}:${port1}`, '-p', `${port2}:${port2}`,
               '-p', `${port3}:${port3}`, '-p', `${port4}:${port4}`,
       ]
      runOptions && runOptions.length != 0 && cmd.push(`${runOptions}`)
      cmd.push(this._settings.rnodeDockerImage)
      cmd.push(...runCmd('/vscode/.rnode'))
      return cmd
    }

    // Start RNode (standalone) process used by the server
    let vm: ChildProcess
    if (this.useDocker()) {
      vm = spawn('docker', dockerCmd(this._settings.dockerRunOptions))
    } else {
      // RNode run
      vm = spawn(this._settings.rnode, runCmd(dataDir))
    }

    if (!vm.pid) {
      vm.on('error', error => {
        this.log(`${error}`)
      })
      return
    }

    vm.stderr.on('data', data => {
      const result = `${data}`

      if (result.trim() === '') return

      // Display error on last evaluated editor document
      // - until all errors will be returned to the caller
      // - https://github.com/rchain/rchain/pull/2143
      // Sample: Syntax Error, trying to recover and continue parse... for input symbol "" spanning from unknown:15/24(460) to unknown:15/25(461)
      const syntaxErrorRegex = /Syntax Error[\s\S]+spanning from unknown:(\d+)\/(\d+)\(\d+\) to unknown:(\d+)\/(\d+)\(\d+\)/g
      const syntaxError = regexec(syntaxErrorRegex, result)
      if (syntaxError) {
        const [message, ...ranges] = syntaxError
        const [sl, sc, el, ec] = ranges.map(x => parseInt(x) - 1)
        const range = {
          start: { line: sl, character: sc },
          end: { line: el, character: ec },
        }
        this.sendErrors(this._uri, {range, message})
        this.log('')
        this.log(`Error: ${message}`)
      }

      // [rchain] Error: ...
      const rchainError = regexec(/\[rchain\] (Error: .+)./g, result)
      if (rchainError) {
        const [_, err] = rchainError
        this.log('')
        this.log(err)
      }

      const knownMsg = syntaxError || rchainError
      if (!knownMsg) {
        // This will catch Docker downloading messages
        process.stdout.write(data)
      }
    })

    let isStarted = false

    // RNode stdout
    vm.stdout.on('data', data => {
      const result = `${data}`.replace(/[\n\r]*$/, '').trim()

      if (result === '') return

      // Display RNode output
      if (this._settings.showAllOutput) {
        this.log(result)
      } else if (!this.isReady) {
        // Display RNode version info
        // RChain Node 0.9.1.git4af26be5 (4af26be585c905097be5d474acce693e059ab64b)
        const versionInfo = !isStarted && regexec(/INFO  coop.rchain.node.Main\$ - (RChain Node .*)/g, result)
        if (versionInfo) {
          // Display version info
          const [_, version] = versionInfo
          process.stdout.write(`Starting ${version} `)
          isStarted = true
        }
        // Display progress as dots
        isStarted && process.stdout.write('.')
      } else {
        // Display evaluated code and result (skip info/warn logs)
        const nodeLog = regexec(/\[[a-z-]+[\d]+\] (WARN|INFO)/g, result)
        !nodeLog && this.log(result)
      }

      // Check for log message that node is ready
      const startListen = regexec(/Listening for traffic on rnode:\/\/([\s\S\d]+)./g, result)
      if (startListen) {
        // Create new gRPC client (after losing connection)
        this.grpcClient = new grpc.Client(rnodeUrl, grpc.credentials.createInsecure())
        // Get repl service path
        const replPath = 'coop.rchain.node.model.Repl'
        rnodeGetReplService(this.grpcClient)
          .then(({name}) => name)
          .catch(err => {
            this.log(``)
            this.log(`Can\'t get Repl service name, fallback to default ${replPath}, ${err.message}`)
            return replPath
          })
          .then(name => this.grpcReplServicePath = name)
          .finally(() => {
            this.isReady = true
            const [_, address] = startListen
            const msgReady = `Rholang VM is ready!`
            const msgListen = `Listening on rnode://${address}`
            const msg = `${msgReady} ${msgListen}`
            this.log('')
            this.log(msgListen)
            this.log(`Ports - gRPC external/internal ${port1}/${port2}`)
            this.log(`      - rnode/http/kademlia ${port0}/${port3}/${port4}`)
            this.log(msgReady)
            this.connection.window.showInformationMessage(msg)
          })
      }
    })
    return vm
  }

  evalCode(uri: string) {
    // Global uri until caller gets all errors
    this._uri = uri
    const document = this.documents.get(uri)
    const codeStringRaw = document.getText().replace(/\r/g, '')
    const codeString    = this._settings.unsupportedNamesRemoveEnable
      ? this.removeUnsupportedNames(codeStringRaw)
      : codeStringRaw

    // Start Rholang VM (eval)
    this.log('')
    this.log('>>>>>>>>>>>>>>>>>>>>>> Sending code to Rholang VM >>>>>>>>>>>>>>>>>>>>>>')

    // Evaluate editor text
    rnodeReplEval({program: codeString, printunmatchedsendsonly: true}, this.grpcReplServicePath, this.grpcClient)
      .then(({output}) => this.processEvalResult(uri, output))
  }

  // Erase blockchain specific names from the input Rholang code
  removeUnsupportedNames(code: string) {
    const names = this._settings.unsupportedNames.split(' ').map(x => x.trim())
    const removeNames = R.reduce((acc: string, name: string) => {
      const regex = new RegExp(`\\(\`${name}\`\\)`, 'g')
      return acc.replace(regex, `/*${name}*/`)
    })
    return removeNames(code, names)
  }

  processEvalResult(uri, output: string) {
    // Ensure it's string
    const result = `${output}`

    if (result.trim() === '') return

    // Deployment cost: CostAccount(39,Cost(1165))
    // Deployment cost: Cost(718,)                     - v0.9.1
    const costRegex = /(Deployment cost:[^\n]+)/g
    const cost = regexec(costRegex, result)
    if (cost) {
      const [_, costMsg] = cost
      this.sendClear(uri)
      this.log('')
      this._settings.showAllOutput
        ? this.log(result)
        : this.log(costMsg)
    }

    // Compile errors

    const errorParsers = [{
      // Free names
      // Error: coop.rchain.rholang.interpreter.errors$TopLevelFreeVariablesNotAllowedError: Top level free variables are not allowed: x at 18:5.
      regex: /Error:[\s\S]+ (\d+):(\d+)\./g,
      diag: ([message, ...ranges]) => {
        const [sl, sc] = ranges.map(x => parseInt(x) - 1)
        const range = {
          start: { line: sl, character: sc },
          end: { line: sl, character: sc },
        }
        return [{ range, message }]
      }
    }, {
      // Name/Process type error
      // Error: coop.rchain.rholang.interpreter.errors$UnexpectedProcContext: Name variable: x at 17:5 used in process context at 17:17
      // Error: coop.rchain.rholang.interpreter.errors$UnexpectedNameContext: Proc variable: x at 19:6 used in Name context at 19:19
      regex: /Error:[\s\S]+ (\d+):(\d+) used in \w+ context at (\d+):(\d+)/g,
      diag: ([message, ...ranges]) => {
        const [sl1, sc1, sl2, sc2] = ranges.map(x => parseInt(x) - 1)
        // Make range with the same start/end
        const r = (line, character, msg) => ({ range: { start: { line, character }, end: { line, character }, }, message: msg })
        return [r(sl1, sc1, message), r(sl2, sc2, message)]
      }
    }, {
      // Error with line identification
      // Error: coop.rchain.rholang.interpreter.errors$LexerError: Illegal Character <?> at 4:4(9)
      regex: /Error:[\s\S]+ (\d+):(\d+)[\s\S]+/g,
      diag: ([message, ...ranges]) => {
        const [sl, sc] = ranges.map(x => parseInt(x) - 1)
        const range = {
          start: { line: sl, character: sc },
          end: { line: sl, character: sc },
        }
        return [{ range, message }]
      }
    }, {
      // Errors received during evaluation
      regex: /Errors received during evaluation:[\n\r]*([\s\S\n\r]+)Storage Contents:/g,
      diag: ([message, error]) => {
        const range = {
          start: { line: 0, character: 0 },
          end: { line: Number.MAX_VALUE, character: Number.MAX_VALUE },
        }
        return [{ range, message: `Error: ${error}` }]
      }
    }, {
      // Any other error
      regex: /Error:[\s\S]+/g,
      diag: ([message, ...ranges]) => {
        const [sl, sc] = ranges.map(x => parseInt(x) - 1)
        const range = {
          start: { line: 0, character: 0 },
          end: { line: Number.MAX_VALUE, character: Number.MAX_VALUE },
        }
        return [{ range, message }]
      }
    }]

    const showError = str => ({regex, diag}) => {
      const match = regexec(regex, str)
      if (match) {
        const errors = diag(match)
        this.sendErrors(uri, ...errors)
        this.log('')
        this.log(errors[0].message)
      }
      return match
    }

    // Detect syntax error printed on RNode stdout
    const syntaxError = regexec(/Error:[\s\S]+ Unrecognized interpreter error/g, result)

    const shownMsg = syntaxError || errorParsers.find(showError(result)) || cost
    if (this._settings.showAllOutput && !shownMsg) {
      this.log('')
      this.log(`EVAL RESULT:\n${result}`)
    }
  }

}
